### PR TITLE
Fix escape unescape text

### PR DESCRIPTION
**Pull request by islemaster**
I found an edge case in the `escapeText` and `unescapeText` helpers we use inside App Lab's `getText` and `setText` methods.  The following round-trip failed:

``` javascript
"\nLine2\nLine3" === unescapeText(escapeText("\nLine2\nLine3"));
// Actual result: "Line2\nLine3"
```

We lose the leading newline in a round-trip operation.  Oddly enough, the multiple-leading-newlines cases was working but having a single leading newline broke.
# Repro Steps

You can repro this issue in live App Lab (until it gets fixed by following these directions):
1. Go to https://studio.code.org/projects/applab/oGZfD95wvpD-kBy_DoO3yA
1. Click in the textarea
1. Type {Enter}Line2
1. Hold shift and press the up arrow, to select both lines
1. Press backspace to delete the content so far
1. Type {Enter}Line2 a second time
1. Click the button

Expected: The text does not change (running setText(getText()))

Actual: You lose the first line
# Root Cause

If you examine the textarea contents at each of the steps above, the first time you type `{Enter}Line2` you get `<div><br></div><div>Line2</div>`.  When this is the DOM content, the round-trip works correctly.

The second time you type that same thing, you end up with `<br><div>Line2</div>`.  Our round-trip isn't handling this case correctly - when I originally wrote these transforms, I didn't account for the case where the first line is nonempty but only contains a tag.
# The fix

I remove a gross "strip leading nondoubled newline" hack I was using to fix this problem, and instead I replace every div with a newline _except_ a div that is the very beginning of content.
# The tests

```
  utils.escapeText
    ✓ no-op on empty string
    ✓ no-op on alphanumeric string
    ✓ replaces all & with &amp;
    ✓ replaces all < with &lt;
    ✓ replaces all > with &gt;
    ✓ breaks up all doubled spaces with &nbsp;
    ✓ Escapes already-escaped content
    ✓ Wraps second and subsequent lines in <div>, removing line breaks
    ✓ Uses <br> for blank first line when handling multiple lines
    ✓ Adds <br> to every other blank line

  utils.unescapeText
    ✓ no-op on empty string
    ✓ no-op on alphanumeric string
    ✓ replaces all &amp; with &
    ✓ replaces all &lt; with <
    ✓ replaces all &gt; with >
    ✓ replaces all &nbsp; with spaces
    ✓ Unescapes pre-escaped content correctly
    ✓ Unwraps <div>-wrapped lines, adding line breaks
    ✓ If input starts with <div> treats that as the first line
    ✓ Converts <div><br></div> to a blank line
    is the inverse of escapeText
      ✓ With newline
      ✓ With leading newline
      ✓ With multiple leading newlines
      fuzz tests (randomized)
        ✓ <div>&nbsp;&gt;&amp;<&gt;&amp;&nbsp;<&lt;
        ✓ &</div>>&lt;>\n>\n&gt;&
        ✓ <&nbsp;>\n<div>\n<a><a>&&nbsp;
        ✓ <a>\n&lt;<br><div><div><a><br>\n&lt;
        ✓ &nbsp; </div>> <br><div>\n&gt; 
        ✓ &nbsp;&&<div><a>&gt;&amp; </div><br>
        ✓ </div></div>&lt;&amp;<a>&nbsp;</div><div><a>>
        ✓ < <br>&nbsp;<br><br>&lt;<div> &nbsp;
        ✓ >&&lt;&gt;&lt;<&lt;&gt;&nbsp;<a>
        ✓ &amp;>> <a>\n>&nbsp;><a>


  33 passing (424ms)
```

These methods were tested via the getText/setText tests in applab, but they weren't tested in isolation (they were extracted to utils since originally written) so I added some direct utility tests around them that are more robust.

### CODE REVIEW

File: apps/src/utils.js [hunk 1]
```diff
@@ -331,28 +331,38 @@ exports.browserSupportsCssMedia = function () {
...
+  //
+  //     Line 2
+  //   Renders as any of:
+  //    <br><div>Line 2</div>
+  //    <div><br></div><div>Line 2</div>
 
+  // First, convert every <div> tag that isn't at the very beginning of the string
+  // to a newline.  This avoids generating an incorrect blank line at the start
+  // if the first line is wrapped in a <div>.
+  cleanedText = cleanedText.replace(/(?!^)<div>/gi, '\n');
```
Comment (islemaster) [inline apps/src/utils.js:363]: Note: `(?!^)` is a regex negative assertion - in this case, it means "and we can't match the beginning of the text here."

File: apps/test/applabTest.js [hunk 2]
```diff
@@ -174,7 +174,7 @@ describe('getText/setText commands', function () {
 
       it('does add leading newline for leading empty div', function () {
         element.innerHTML = '<div><br></div><div>text</div><div>with</div><div>leading empty div</div>';
-        assert.equal(getInnerText(element), '\n\ntext\nwith\nleading empty div');
+        assert.equal(getInnerText(element), '\ntext\nwith\nleading empty div');
```
Comment (islemaster) [inline apps/test/applabTest.js:177]: I think this was an actual bug in the existing tests - if you expand and look at the test above this one it checks 'no leading newline' expecting `text\nwith...` - so I'm not sure why this 'with leading newline' test expected two leading newlines `\n\ntext\nwith...`

File: apps/test/utilityTests.js [hunk 3]
```diff
@@ -449,3 +449,258 @@ describe('forceInsertTopBlock', function () {
...
+        '<a>',
+        '&',
+        '<',
+        '>',
+        ' ',
+        '&amp;',
+        '&lt;',
+        '&gt;',
+        '&nbsp;',
+        '\n'
```
Comment (islemaster) [inline apps/test/utilityTests.js:682]: I'm open to suggestions here - I just threw in components that I know are likely to be problematic for this transformation.

**Comment by joshlory** [comment]
LGTM :+1: Good tests!

**Pull request merged**